### PR TITLE
[JavaWorker]Changes to the directory under src for support java worker

### PR DESCRIPTION
**Pull request by salah-man**
--------------------------
This commit includes changes to the directory under src, which is part of the java worker support of Ray.
It consists of the following changes:
 src/common/task.cc - just fix null point problem
 org_ray_spi_impl_DefaultLocalSchedulerClient.* - JNI support for local scheduler client, and the org_ray_spi_impl_DefaultLocalSchedulerClient.cc file is not autogenerated

### CODE REVIEW

**Comment by robertnishihara** [comment]
#2095

File: src/common/task.cc [hunk 1]
```diff
@@ -324,7 +324,10 @@ int TaskSpec_arg_id_count(TaskSpec *spec, int64_t arg_index) {
   RAY_CHECK(spec);
   auto message = flatbuffers::GetRoot<TaskInfo>(spec);
   auto ids = message->args()->Get(arg_index)->object_ids();
-  return ids->size();
+  if (ids == nullptr)
+    return 0;
+  else
+    return ids->size();
```
Comment (robertnishihara) [inline src/common/task.cc]: The linting doesn't seem to be enforcing this, but please change this to

```c++
if (ids == nullptr) {
  return 0;
} else {
  return ids->size();
}
```

File: src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc [hunk 2]
```diff
@@ -0,0 +1,224 @@
+
```
Comment (robertnishihara) [inline src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc]: remove newline at beginning

File: src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+ public:
+  UniqueID *PID;
+
+  UniqueIdFromJByteArray(JNIEnv *env, jbyteArray wid) {
+    _env = env;
+    _bytes = wid;
+
+    jbyte *b = (jbyte *) _env->GetByteArrayElements(_bytes, NULL);
+    PID = (UniqueID *) b;
```
Comment (robertnishihara) [inline src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc]: In this file, let's remove all C style casts and instead use `static_cast`

File: src/common/task.cc (hunk 1 above)
Comment (salah-man) [reply src/common/task.cc]: Got it, thanks for your time, i will change them soon

**Comment by salah-man** [comment]
Excuse me, but i just fix the problem as the comment before, could you review this pr again? Thank you very much. @robertnishihara

File: src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+                                                         jbyteArray actorId,
+                                                         jboolean isWorker,
+                                                         jlong numGpus) {
+  // 	native private static long _init(String localSchedulerSocket,
+  //     byte[] workerId, byte[] actorId, boolean isWorker, long numGpus);
+  UniqueIdFromJByteArray w(env, wid), a(env, actorId);
+  const char *nativeString = env->GetStringUTFChars(sockName, JNI_FALSE);
+  auto client = LocalSchedulerConnection_init(nativeString, *w.PID, isWorker);
+  env->ReleaseStringUTFChars(sockName, nativeString);
+  return (jlong) client;
```
Comment (robertnishihara) [inline src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc]: use C++ cast

File: src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+  local_scheduler_submit(client, taskExecutionSpec);
+}
+
+/*
+ * Class:     org_ray_spi_impl_DefaultLocalSchedulerClient
+ * Method:    _getTaskTodo
+ * Signature: (J)[B
+ */
+JNIEXPORT jbyteArray JNICALL
+Java_org_ray_spi_impl_DefaultLocalSchedulerClient__1getTaskTodo(JNIEnv *env,
```
Comment (robertnishihara) [inline src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc]: I'm a bit confused about where these function names come from. E.g., what does the `1` mean and the `Todo`?

File: src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    jint pos,
+    jint sz) {
+  // task -> TaskInfo (with FlatBuffer)
+  // native private static void _submitTask(long client, /*Direct*/ByteBuffer
+  // task);
+  auto client = reinterpret_cast<LocalSchedulerConnection *>(c);
+  TaskSpec *task =
+      reinterpret_cast<char *>(env->GetDirectBufferAddress(buff)) + pos;
+  std::vector<ObjectID> execution_dependencies;
+  if (cursorId != NULL) {
```
Comment (robertnishihara) [inline src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc]: In general, we should avoid using `NULL` and instead use `nullptr` for pointers. However, I haven't worked much with JNI so should it be different in this case?

File: src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+Java_org_ray_spi_impl_DefaultLocalSchedulerClient__1init(JNIEnv *env,
+                                                         jclass,
+                                                         jstring sockName,
+                                                         jbyteArray wid,
+                                                         jbyteArray actorId,
+                                                         jboolean isWorker,
+                                                         jlong numGpus) {
+  // 	native private static long _init(String localSchedulerSocket,
+  //     byte[] workerId, byte[] actorId, boolean isWorker, long numGpus);
+  UniqueIdFromJByteArray w(env, wid), a(env, actorId);
```
Comment (robertnishihara) [inline src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc]: In general I think we should be using much more informative variable names (e.g., full words) and avoiding abbreviations

File: src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+  auto client = reinterpret_cast<LocalSchedulerConnection *>(c);
+  int64_t task_size = 0;
+
+  // TODO: handle actor failure later
+  TaskSpec *spec = local_scheduler_get_task(client, &task_size);
+
+  jbyteArray result;
+  result = env->NewByteArray(task_size);
+  if (result == NULL) {
+    return NULL; /* out of memory error thrown */
```
Comment (robertnishihara) [inline src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc]: Is the plan to let the caller check for this and throw a Java exception?

**Review by robertnishihara — COMMENTED** [review]
Looks good to me! I left a few more comments and some questions. Thanks!

File: src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc (hunk 4 above)
Comment (salah-man) [reply src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc]: ok

File: src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc (hunk 5 above)
Comment (salah-man) [reply src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc]: Oh, it is auto generated by javah, which make the function name looks stange.

File: src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc (hunk 6 above)
Comment (salah-man) [reply src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc]: I just fix them, it is pointer here

File: src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc (hunk 8 above)
Comment (salah-man) [reply src/local_scheduler/lib/java/org_ray_spi_impl_DefaultLocalSchedulerClient.cc]: yes, currently this situation will throw a Exception in java

**Review by robertnishihara — APPROVED** [review]

**Pull request merged**